src/metadata/authorization_server.rs

```diff
@@ -0,0 +1,155 @@
+            additional_fields: Default::default(),
+        }
+    }
+
+    field_getters_setters![
+        pub self [self] ["authorization server metadata value"] {
+            set_issuer -> issuer[IssuerUrl],
+            set_authorization_endpoint -> authorization_endpoint[Option<AuthUrl>],
+            set_token_endpoint -> token_endpoint[TokenUrl],
+            set_jwks_uri -> jwks_uri[Option<JsonWebKeySetUrl>],
+            set_registration_endpoint -> registration_endpoint[Option<RegistrationUrl>],
+            set_scopes_supported -> scopes_supported[Option<Vec<Scope>>],
+            set_response_types_supported -> response_types_supported[Option<Vec<ResponseType>>],
+            set_response_modes_supported -> response_modes_supported[ResponseModes],
+            set_grant_types_supported -> grant_types_supported[GrantTypesSupported],
+            set_revocation_endpoint -> revocation_endpoint[Option<RevocationUrl>],
+            set_introspection_endpoint -> introspection_endpoint[Option<IntrospectionUrl>],
+            set_code_challenge_methods_supported -> code_challenge_methods_supported[Option<Vec<PkceCodeChallengeMethod>>],
+            set_pre_authorized_grant_anonymous_access_supported -> pre_authorized_grant_anonymous_access_supported[bool],
+            set_pushed_authorization_request_endpoint -> pushed_authorization_request_endpoint[Option<ParUrl>],
+            set_require_pushed_authorization_requests -> require_pushed_authorization_requests[bool],
+        }
+    ];
+
+    pub fn additional_fields(&self) -> &Map<String, Json> {
+        &self.additional_fields
+    }
+
+    pub fn additional_fields_mut(&mut self) -> &mut Map<String, Json> {
+        &mut self.additional_fields
+    }
+}
+
+impl MetadataDiscovery for AuthorizationServerMetadata {
+    const METADATA_URL_SUFFIX: &'static str = ".well-known/oauth-authorization-server";
+
+    fn validate(&self, issuer: &IssuerUrl) -> Result<()> {
+        if self.issuer() != issuer {
+            bail!(
+                "unexpected issuer URI `{}` (expected `{}`)",
+                self.issuer().as_str(),
+                issuer.as_str()
+            )
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct ResponseModes(pub Vec<ResponseMode>);
+
+impl Default for ResponseModes {
+    fn default() -> Self {
+        Self(vec![
+            ResponseMode::new("query".to_owned()),
+            ResponseMode::new("fragment".to_owned()),
+        ])
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct GrantTypesSupported(pub Vec<GrantType>);
+
+impl Default for GrantTypesSupported {
+    fn default() -> Self {
+        Self(vec![GrantType::AuthorizationCode, GrantType::Implicit])
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "snake_case")]
+pub enum GrantType {
+    AuthorizationCode,
+    Implicit,
```

Comment:
Is implicit flow possible, or is it just because it's part of another RFC?

Reply:
It's not possible, I'm just adding it as it is one of the default scopes as per RFC 8414.